### PR TITLE
Machine learning model to predict failing or healthy cell type

### PR DESCRIPTION
**Pull request by jenna-tomkinson**
# Machine learning model to predict failing or healthy cell type

In this PR, I trained a logistic regression model on downsampled Plate 4 training data to predict cell type (failing or healthy). I include three notebooks where I: (1) split the data, (2) train the models, and (3) evaluate the model and output figures.

Some small edits were made to other modules, including outputting figures from single-cell QC, updating the main env, etc.

### CODE REVIEW

**Review by MikeLippincott — APPROVED** [review]
LGTM. I left a few comments and questions to consider but is ready to merge as is.

File: 5.machine_learning/scripts/0.split_data.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+# 1. Remove all wells from DMSO treated healthy heart #7 and remove all wells from one failing heart (random)
+# 2. Remove one well from each heart (both failing and healthy)
+
+# ### Add DMSO treated heart 7 cells to holdout df
+
+# In[4]:
+
+
+# Copy all DMSO heart #7 rows into the holdout_df
+holdout_df = plate_4_df[(plate_4_df['Metadata_heart_number'] == 7) & (plate_4_df['Metadata_treatment'] == 'DMSO')]
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply 5.machine_learning/scripts/0.split_data.py:86]: We only have two healthy hearts, and we decided to include the DMSO treatment as hold-out, which is only applied to heart #7.

File: 5.machine_learning/scripts/0.split_data.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+random_state = 0
+
+# Set the ratio of the test data to 30% (training data will be 70%)
+test_ratio = 0.30
+
+# Split the plate 4 data into training and test
+training_data, testing_data = train_test_split(
+    plate_4_df,
+    test_size=test_ratio,
+    stratify=plate_4_df[["Metadata_cell_type"]],
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply 5.machine_learning/scripts/0.split_data.py:217]: That is a very good point. I think that based on the performance of the hold-out data we do not see that heart # as an influence to the prediction.

I will keep this in mind for the future though, since I didn't consider this before. I am curious to see if there is a struggle when it comes to apply to different plates with different heart #s.

File: 5.machine_learning/scripts/1.train_models.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+from sklearn.model_selection import RandomizedSearchCV, StratifiedKFold
+from sklearn.utils import parallel_backend
+from sklearn.exceptions import ConvergenceWarning
+from sklearn.preprocessing import LabelEncoder
+from joblib import dump
+
+import sys
+
+sys.path.append("../utils")
+from training_utils import downsample_data, get_X_y_data
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply 5.machine_learning/scripts/1.train_models.py:26]: When you get a chance, I would love to discuss pre-commit implementation. I will go back and reorder this myself for PEP8 organization.

File: 5.machine_learning/scripts/1.train_models.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+# 
+# **Note:** We will be using RandomizedSearchCV to hyperparameterize the model since we have a larger dataset and it will be easier to try random combinations than all combinations.
+
+# ### Set up the model and hyper parameter method
+
+# In[7]:
+
+
+# Set folds for k-fold cross validation (default is 5)
+straified_k_folds = StratifiedKFold(n_splits=10, shuffle=False)
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply 5.machine_learning/scripts/1.train_models.py:122]: After discussing it with you in person, I have added to my README with more info on why I think this choice, which I implemented from Roshan's phenotypic model, was good to implement with CFReT.

File: 5.machine_learning/scripts/2.evalute_models.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+# Initialize empty lists to store data for each iteration
+precision_list = []
+recall_list = []
+threshold_list = []
+model_type_list = []
+data_type_list = []
+
+for model_path in models_dir.iterdir():
+    print("Evaluating", model_path.stem.split("_")[5], "model...")
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply 5.machine_learning/scripts/2.evalute_models.py:61]: Understandable to try and avoid hardcoding. In this case, I could use a dictionary to assign paths to models with the model type (e.g., shuffled or final) to avoid the hardcode, but that would take some time to refactor the structure. I will keep this in mind for the future, so thanks! 😄

File: 5.machine_learning/README.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
+# 5. Machine learning
+
+In this module, we train a logistic regression elasticnet machine learning model to predict if a single cell is either failing or healthy based on the feature selected morphology features from Plate 4. 
+
```
[the remark this answers is not in the document]
Comment (jenna-tomkinson) [reply 5.machine_learning/README.md:4]: I want to give Roshan's README credit as inspo here for the model description part since I wanted to model the level of detail. Thank you for the compliment since I wanted to make sure I described the evaluation and splitting too!

I think I will definitely add a short description of the data and maybe use a table too!
